### PR TITLE
Sync and search feature implemented

### PR DESCRIPTION
**Pull request by SeraYang1**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,3 +7,24 @@ Three tools to ease the life of Open Source developers:
 - An analytics dashboard (stalled PRs & discussions, time since last comment,
   etc.)
 
+## Command Line
+<b> ```git hub pr _```</b> <br />
+&nbsp;&nbsp; Pulls down and checkout the branch of the pr. <br />
```
Comment (stefanv) [inline README.md]: Odd characters?

File: git-hub [hunk 2]
```diff
@@ -6,35 +6,27 @@ import click
...
-    "git checkout -b pr/num user/branch"
-
-    Parameters
-    ----------
-    num : integer
-        The number of the pull request.
-    """
-    process = subprocess.Popen("git remote -v", stdout=subprocess.PIPE)
+def login():
+    process = subprocess.Popen(["git","remote","-v"], stdout=subprocess.PIPE)
```
Comment (stefanv) [inline git-hub]: Please run all code through `flake8` for style checking.

File: git-hub [hunk 3]
```diff
@@ -43,16 +35,35 @@ def pr(num):
...
         pr = g.get_user(username).get_repo(repo).get_pull(num)
         label = pr.head.label
         other_user, branch = label.split(":")
+        process = subprocess.Popen(["git","remote","-v"], stdout=subprocess.PIPE)
+        remotes = str(process.stdout.read())
         remote_name = "\\n" + other_user + "\\t"
         if remote_name not in remotes:
-            p = subprocess.Popen(["git", "remote", "add", other_user,
-                                 f'git@github.com:{other_user}/{repo}'])
+            print(other_user)
```
Comment (stefanv) [inline git-hub]: Debug statements left behind

File: git-hub [hunk 4]
```diff
@@ -87,14 +97,142 @@ def push():
...
+            p = subprocess.Popen(["git", "push", other_user, f'{pr}:{branch}'], stdout=subprocess.PIPE)
             p.stdout.read()
     except FileNotFoundError as e:
         print(e)
         print("This command must be used after git hub pr.")
         sys.exit(1)
 
+def search(keyword, user, comment, number, branch):
+    """Searches open and closed pull request comments for specified keyword.
+    Opens pull-requests.toml file in .git folder to fetch pull requests.
```
Comment (stefanv) [inline git-hub]: Blank line after first line of docstring.  (1 line for a short description, then a second paragraph with longer description.)

File: git-hub [hunk 5]
```diff
@@ -87,14 +97,142 @@ def push():
...
+        Search by PR comment
+    number : string
+        Search by PR number
+    branch : string
+        Search by PR branch
+    """
+    try:
+        open_prs = False
+        closed_prs = False
+        pathToGit = './'
```
Comment (stefanv) [inline git-hub]: path_to_git

File: git-hub [hunk 6]
```diff
@@ -87,14 +97,142 @@ def push():
...
+        while not any(fname == '.git' for fname in os.listdir(pathToGit)):  #if not in directory with .git, keep going back to find file
```
Comment (stefanv) [inline git-hub]: Use glob.glob

Comment (stefanv) [reply git-hub]: Perhaps walkpath?

File: git-hub [hunk 7]
```diff
@@ -87,14 +97,142 @@ def push():
...
+            pathToGit = pathToGit+"../"
```
Comment (stefanv) [inline git-hub]: Where would this loop end?  It could continue forever!

File: git-hub [hunk 8]
```diff
@@ -87,14 +97,142 @@ def push():
...
+        pathToGit = pathToGit+".git/git-hub/pull-requests.toml"
+        f = open(pathToGit,"r")
+        pr_dict = toml.load(f)      #fetches toml file and creates a dictionary
+        open_dict = pr_dict['open pull requests']
+        closed_dict = pr_dict['closed pull requests']
+
+        for key in open_dict.keys():        #iterates through and checks if open pull requsts have keyword
+            pull = open_dict[key]
+            if keyword:         #checks keyword
+                if keyword.upper() in pull.upper() or keyword in str(key):
```
Comment (stefanv) [inline git-hub]: also case insensitive matching on key?

File: git-hub [hunk 9]
```diff
@@ -87,14 +97,142 @@ def push():
...
+                    if not open_prs:
+                        click.echo("OPEN pull requests")
+                        open_prs = True
+                    click.echo('{0}  {1}'.format(key, pull))
+            else:               #checks for all other parameters
+                tempUser, tempBranch = pull.split('/', 1)
```
Comment (stefanv) [inline git-hub]: temp_user, temp_branch, etc.

File: git-hub [hunk 10]
```diff
@@ -87,14 +97,142 @@ def push():
...
+                tempBranch, tempComment = tempBranch.split(': ',1)
+                if number in str(key) and comment.upper() in tempComment.upper() and user.upper() in tempUser.upper() and branch.upper() in tempBranch.upper():
+                    if not open_prs:
+                        click.echo("OPEN pull requests")
+                        open_prs = True
+                    click.echo('{0}  {1}'.format(key, pull))
+
+
```
Comment (stefanv) [inline git-hub]: Extra blank lines

File: git-hub [hunk 11]
```diff
@@ -87,14 +97,142 @@ def push():
...
+
+        for key in closed_dict.keys():
```
Comment (stefanv) [inline git-hub]: for key in closed_dict:

File: git-hub [hunk 12]
```diff
@@ -87,14 +97,142 @@ def push():
...
+                    if not closed_prs:
+                        if open_prs:
+                            click.echo()
+                            click.echo("--------------------------------")
+                            click.echo()
+                        click.echo("CLOSED pull requests")
+                        closed_prs = True
+                    click.echo('{0}  {1}'.format(key, pull))
+            else:
+                tempUser, tempBranch = pull.split('/', 1)
```
Comment (stefanv) [inline git-hub]: Don't repeat code; refactor the few lines above into a function, and re-use here.

**Review by stefanv — COMMENTED** [review]
Thanks, Sera, this looks good!  I left a few initial comments, and will have another look after you incorporated those changes.

File: README.md (hunk 1 above)
Comment (SeraYang1) [reply README.md]: That's supposed to be a space, I'm not sure how to indent or tab on markdown files besides adding spaces. Is there a better way to go about doing this?

File: git-hub (hunk 6 above)
Comment (SeraYang1) [reply git-hub]: Sorry, not sure I understand what you mean, could you elaborate?

File: git-hub (hunk 8 above)
Comment (SeraYang1) [reply git-hub]: should I make it case sensitive?

File: README.md (hunk 1 above)
Comment (stefanv) [reply README.md]: How about a list:

```
- `git hub pr -` :  Description goes here.
```

Comment (SeraYang1) [reply README.md]: oh that looks much better! will change. thanks for the tip!

File: README.md [hunk 13]
```diff
@@ -7,3 +7,8 @@ Three tools to ease the life of Open Source developers:
 - An analytics dashboard (stalled PRs & discussions, time since last comment,
   etc.)
 
+## Command Line
+- `git hub pr -` :  Pulls down and checkout the branch of the pr. 
+- `git hub push` :  Pushes changes back to a branch.
+- `git hub sync` :  Updates and saves pull-requests in pull-requests.toml in the .git folder.
+- `git hub search -` :  Searches in pull-requests.toml to find pull-requests that match search keywords.
```
Comment (stefanv) [inline README.md]: We want to hide any internal implementation detail from the user.

File: git-hub [hunk 14]
```diff
@@ -6,35 +6,33 @@ import click
 import sys
 import textwrap
 import json
+import toml
 import os
+import queue as Q
```
Comment (stefanv) [inline git-hub]: Rather just import queue

File: git-hub [hunk 15]
```diff
@@ -87,15 +103,292 @@ def push():
...
+
+def print_in_order(sort, q):
+    """
+    Prints items in q sorted by order given by sort
+
+    Parameters
+    ----------
+    sort : string
+        sort in increasing or decreasing order
+    q : PriorityQueue
```
Comment (stefanv) [inline git-hub:194]: Why not just sort a list?  Sort can take a function argument that determines which keys are used for sorting.

File: git-hub [hunk 16]
```diff
@@ -87,15 +103,292 @@ def push():
...
+    else:
+        stack = []
+        while not q.empty():
+            pair = q.get()
+            stack.append("{0} : {1}".format(pair[1], pair[0][:10]))
+        while not len(stack) == 0:
+            click.echo(stack.pop())
+
+
+def findMatch(open_or_closed, list_of_dictionaries, keyword, user, comment, number, branch, label, appeared_before, line):
```
Comment (stefanv) [inline git-hub]: Please run all code through `flake8`.

File: git-hub [hunk 17]
```diff
@@ -87,15 +103,292 @@ def push():
...
+    number : string
+        Search by PR number
+    branch : string
+        Search by PR branch
+    appeared_before : boolean
+        Whether the search match is the first match
+    line : boolean
+        Whether to draw lines before listing closed pull requests
+    """
+    q = Q.PriorityQueue()
```
Comment (stefanv) [inline git-hub]: I think we can replace the content of this function with a generic dictionary matching function.  Something like:

`filter_dictionary(PRs, constraints)`

where `PRs` is a list of PR entries, and `constraints` is a dictionary.

File: git-hub [hunk 18]
```diff
@@ -87,15 +103,292 @@ def push():
...
+        path_git = find_path_to_git()
+        path_github = pjoin(path_git, 'git-hub')
+        path_prs = pjoin(path_github, 'pull-requests.toml')
+        f = open(path_prs, "r")
+
+        pr_dict = toml.load(f)      # fetches toml file and creates a dictionary
+        open_dict = pr_dict['open pull requests']
+        closed_dict = pr_dict['closed pull requests']
+
+        if(not opened or opened.lower() == 'open'):
```
Comment (stefanv) [inline git-hub]: This statement is incorrect.

File: git-hub [hunk 19]
```diff
@@ -87,15 +103,292 @@ def push():
...
+            open_prs, q = findMatch(True, open_dict, keyword, user, comment, number, branch, label, False, False)
+            print_in_order(sort, q)
+        if(not opened or opened.lower() == 'closed'):
+            closed_prs, q = findMatch(False, closed_dict, keyword, user, comment, number, branch, label, False, open_prs)
+            print_in_order(sort, q)
+        if not open_prs and not closed_prs:     # outputs if keyword was not contained in pull requets comments
+            click.echo("Could not find in pull requests. Update your pull requests with 'git hub sync' and try again.")
+    except (OSError, IOError) as e:
```
Comment (stefanv) [inline git-hub]: Only put the try except clause around the minimal amount of code possible.

File: git-hub [hunk 20]
```diff
@@ -87,15 +103,292 @@ def push():
...
+        # if pull-requests.toml hasnt been created yet calls sync and then reties to fetch
+        sync()
+        search(keyword, user, comment, number, branch, status)
+
+
+def parse_to_toml(username, repo, pr, token, open_or_closed):
```
Comment (stefanv) [inline git-hub]: Can we come up with a better naming here?  What is being parsed into toml, and why?

File: git-hub [hunk 21]
```diff
@@ -87,15 +103,292 @@ def push():
...
+
+        to_write = to_write[:-2]+'\n'
+        to_write = to_write + "[['closed pull requests']] \n"
+        closed_prs = g.get_user(username).get_repo(repo).get_pulls("closed")
+        for pr in closed_prs:
+            to_write = to_write + parse_to_toml(username, repo, pr, token, "closed")
+        # converts it to toml and stores in file
+        toml_string = toml.loads(to_write)
+        toml.dump(toml_string, f)
+    except github.BadCredentialsException as e:
```
Comment (stefanv) [inline git-hub]: Again, minimal try excepts.

**Review by stefanv — COMMENTED** [review]
Making good progress!

**Pull request merged**